src/libsystemd/sd-journal/journal-verify.c

```diff
@@ -1216,13 +1232,15 @@ int journal_file_verify(
 
                                 f->hmac_running = false;
                                 last_tag_realtime = rt;
-                                last_sealed_realtime = entry_realtime;
                         }
 
                         last_tag = p + ALIGN64(le64toh(o->object.size));
 #endif
 
                         last_epoch = le64toh(o->tag.epoch);
+                        if (n_tags == 0){
+                                last_epoch--;
```

Comment:
Ooh yes, this is a problem for "legacy" verification. However for continuous verification, the underflow is intended and necessary. Let's see how we can adjust the conditions so we fit all corner cases. Currently the check/access to `last_epoch` are:

- initialized to zero
- when a tag occurs:
  - in "continuous mode", the tag must be the first tag or the epoch must be `last_epoch + 1`
  - in "legacy mode", the epoch must be equal or larger than `last_epoch`
  - in both cases: `last_epoch` is set to the epoch (or one less for continuous mode) 

Journals (immediately after key generation) "usually" start with epoch 0, because the tag for the header is duplicated the only valid sequence in continuous  mode is: 0, 0, 1, 2, 3; and, if rotation happens now, 3, 3, 4, 5, 6 in the next file.

For "legacy" mode, also 0, 0, 1, 5, 6; rotate; 6, 6, 7, 8 is allowed. So maybe we should keep `last_epoch` untouched (i.e. drop line 1242) and make the check more complex:

`! (n_tags == 0 || (n_tags == 1 && le64toh(o->tag.epoch) == last_epoch) || le64toh(o->tag.epoch) == last_epoch + 1)`

of course that condition could be split up if that makes it easier to digest:

```
if(n_tags == 0) {
    // Pass
} else if (n_tags == 1) {
    if (le64toh(o->tag.epoch) != last_epoch) {
        // log error
        goto fail;
    }
} else if (le64toh(o->tag.epoch) == last_epoch + 1) {
    // log error
    goto fail;
}
```

what's your opinion on those two options?

Reply:
Sorry, still I am not familiar with details about tagging. I just commented that for general coding nitpicks. Feel free to take a way you think that should be appropriate here. Hopefully @poettering and/or others will take final review later.